### PR TITLE
Typescript support

### PR DESCRIPTION
**Pull request by s00d**
# Description

rewrote everything to ts

Fixed several possible bugs.

--- 

I hope I rewrote everything to ts for a reason, please rate my work 👍

### CODE REVIEW

**Comment by afshinm** [comment]
@s00d thanks a lot! Looks amazing. I couldn't load the hello world example here https://5fffee77c66089000733881c--introjs-repo.netlify.app/example/hello-world/index.html though

**Comment by s00d** [comment]
Sorry my mistake. Rollup works strangely with multiple imports. Fixed

File: src/IntroJs.ts [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+import {
+  currentStep,
+  goToStep,
+  goToStepNumber,
+  nextStep,
+  previousStep,
+} from "./core/steps";
+
+
+export interface IntroItem {
```
Comment (SamanthaAdrichem) [inline src/IntroJs.ts:25]: I personally love this change to typescript. However I would advice to put each interface in it's own file and let the name end with Interface or start with an extra I, so people know it's not a class but an interface, and they can more easily re-use it outside.

Also checkout this blog article, this way you can compile it as both an ejs and an mjs module
https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html

Comment (s00d) [reply src/IntroJs.ts:25]: This does not make any sense, the typescript create interfaces in the dist folder

**Comment by s00d** [comment]
ann new update

Why do we need a typescript?

js version:
image

ts version:
image

**Comment by joshsny** [comment]
Any updates on getting this merged, looks amazing!

**Comment by mslighthouse** [comment]
This PR is incredible, hope the little conflicts get checked and this can be reviewed again.

**Comment by s00d** [comment]
> Any updates on getting this merged? Looks amazing!

Updated, but the authors do not want to merge (

File: src/core/DOMEvent.ts [hunk 2]
```diff
@@ -35,9 +40,17 @@ const DOMEvent = (() => {
      * @param useCapture Boolean
      * @return null
      */
-    this.on = function (obj, type, listener, context, useCapture) {
+    // @ts-ignore
```
Comment (afshinm) [inline src/core/DOMEvent.ts:43]: These `ts-ignore` statements are kinda concerning. I suppose that's here to disable the typechecking?

**Review by afshinm — COMMENTED** [review]
I appreciate your work and contribution to the project. I still think moving to TS is the right choice but we probably have to do some minor refactoring before changing the setup. I'd like to know what you think.

File: .nvmrc [hunk 3]
```diff
@@ -0,0 +1 @@
+v14.17.1
```
[the remark this answers is not in the document]
Comment (s00d) [reply .nvmrc]: updated

File: src/core/hint.ts [hunk 4]
```diff
@@ -28,8 +31,13 @@ export function hintQuerySelectorAll(selector) {
...
  */
-export async function hideHint(stepId) {
-  const hint = hintQuerySelectorAll(`.introjs-hint[data-step="${stepId}"]`)[0];
+export async function hideHint(
+  this: IntroJs,
+  stepId: number | undefined | string
+) {
+  const hint = hintQuerySelectorAll(
+    `.introjs-hint[data-step="${stepId}"]`
+  )[0] as HTMLElement;
```
[the remark this answers is not in the document]
Comment (s00d) [reply src/core/hint.ts]: refactored

**Comment by boredland-ioki** [comment]
this is incredibly sad.

**Comment by afshinm** [comment]
#1886

**Pull request closed without merging**